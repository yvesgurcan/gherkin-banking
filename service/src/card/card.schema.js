import { gql } from 'apollo-server';

const CreateCardFields = `
    walletId: ID!
    userId: ID!
    balance: Float
    currency: Currency
`;

const CardFields = `
    ${CreateCardFields}
    id: ID
    number: String
    expiration: String
    ccv: String
    isBlocked: Boolean
    createdAt: String
    updatedAt: String

`;

export default gql`
    enum Currency {
        USD
        GBP
        EUR
    }

    type Card {
        ${CardFields}
    }

    type Query {
        cards: [Card]
    }

    type Mutation {
        createCard(${CreateCardFields}): Card
    }
`;
