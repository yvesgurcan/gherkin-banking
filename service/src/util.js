import { ApolloError } from 'apollo-server';
import mongoose from 'mongoose';

// use the findAndModify() function from MongoDB under the hood instead of the findOneAndUpdate() function from Mongoose;
mongoose.set('useFindAndModify', false);

export async function connect() {
    try {
        await mongoose.connect('mongodb://35.164.133.55/gherkin-banking', {
            useNewUrlParser: true,
            useUnifiedTopology: true
        });
    } catch (error) {
        console.error('Could not connect to DB', error);
    }
}

export function close(result) {
    mongoose.connection.close();
    return result;
}

export function parseMongooseErrors(error) {
    if (error.errors) {
        const messages = Object.keys(error.errors).map(key => {
            const { message, path } = error.errors[key];
            return { message, path };
        });
        throw new ApolloError('ValidationError', messages);
    }

    if (error) {
        return error;
    }

    return null;
}

export function generateRandomNumber(length) {
    const max = Number('9'.repeat(length));
    const min = Number('1' + '0'.repeat(length - 1));
    const numberWithinRange = Math.floor(Math.random() * (max - min) + min);
    return numberWithinRange;
}
