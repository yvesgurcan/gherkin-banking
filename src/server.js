import { ApolloServer } from 'apollo-server';
import { mergeSchemas } from 'graphql-tools';
import schemas from './schema.js';
import resolvers from './resolvers.js';

const schema = mergeSchemas({
    schemas,
    resolvers
});

export const serverConfig = {
    schema,
    context: ({ req }) => ({
        userId: req.headers['user-id'],
        companyId: req.headers['company-id']
    })
};

export const instantiateServer = (customConfig = {}) =>
    new ApolloServer({
        ...serverConfig,
        ...customConfig
    });

export async function startServer(server = instantiateServer(), port = '4001') {
    try {
        const { url } = await server.listen({ port });
        console.log(`ðŸš€  Apollo Server ready at ${url}.`);
    } catch (error) {
        console.error('Could not start server.', error);
    }
}
