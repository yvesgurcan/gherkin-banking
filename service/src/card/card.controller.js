import { connect, close, parseMongooseErrors } from '../util';
import CardModel from './card.model.js';

export async function getCards() {
    connect();
    try {
        const result = await CardModel.find();
        close();
        return result;
    } catch (error) {
        close();
        parseMongooseErrors(error);
    }
}

export async function createCard(_, card) {
    connect();
    try {
        const result = await CardModel.create(card);
        close();
        return result;
    } catch (error) {
        close();
        parseMongooseErrors(error);
    }
}
