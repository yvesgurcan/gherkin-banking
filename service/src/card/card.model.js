import mongoose from 'mongoose';
import { ObjectId } from 'mongodb';
import { generateRandomNumber } from '../util';

const cardSchema = new mongoose.Schema(
    {
        walletId: {
            type: ObjectId,
            required: [true, 'Wallet identifier is required.']
        },
        userId: {
            type: ObjectId,
            required: [true, 'User identifier is required.']
        },
        balance: {
            type: Number,
            default: 0
        },
        currency: {
            type: String,
            enum: ['USD', 'GBP', 'EUR'],
            default: 'USD'
        },
        // should be encrypted
        number: {
            type: String,
            default: () => generateRandomNumber(16),
            minlength: 16,
            maxlength: 16
        },
        expiration: {
            type: Date,
            default: new Date() // add 1 month
        },
        ccv: {
            type: String,
            default: () => generateRandomNumber(3),
            minlength: 3,
            maxlength: 3
        },
        isBlocked: {
            type: Boolean,
            default: false
        }
    },
    {
        timestamps: true,
        versionKey: false
    }
);

cardSchema.virtual('id').get(function() {
    return this._id.toHexString();
});

export default mongoose.model('Card', cardSchema);
