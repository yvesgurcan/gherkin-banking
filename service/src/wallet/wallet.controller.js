import { connect, close, parseMongooseErrors } from '../util';
import WalletModel from './wallet.model.js';

export async function getWallets() {
    connect();
    try {
        const result = await WalletModel.find();
        close();
        return result;
    } catch (error) {
        close();
        parseMongooseErrors(error);
    }
}

export async function createWallet(_, wallet) {
    connect();
    try {
        const result = await WalletModel.create(wallet);
        close();
        return result;
    } catch (error) {
        close();
        parseMongooseErrors(error);
    }
}
