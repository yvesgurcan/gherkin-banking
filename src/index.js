import { startServer } from './server';
import { createDatabaseConnection, closeDatabaseConnection } from './database';

startServer();
createDatabaseConnection();

async function exitGracefully() {
    await closeDatabaseConnection();
    process.exit(0);
}

process
    // cleanup when process is terminated
    .on('SIGINT', exitGracefully)
    .on('SIGTERM', exitGracefully)
    // cleanup when process is restarted
    .on('SIGHUP', exitGracefully);
