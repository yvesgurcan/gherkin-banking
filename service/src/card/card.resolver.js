import { getCards, createCard } from './card.controller';

export default {
    Query: {
        cards: getCards
    },
    Mutation: {
        createCard
    }
};
