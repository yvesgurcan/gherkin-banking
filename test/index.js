import { createTestClient } from 'apollo-server-testing';
import { instantiateServer } from '../src/server';
import {
    createDatabaseConnection,
    closeDatabaseConnection,
    dropDatabase
} from '../src/database';
import { GET_WALLETS, CREATE_WALLET } from './fixture/queries.wallet';
import { GET_CARDS, CREATE_CARD } from './fixture/queries.card';

const TEST_DB_NAME = 'gherkin-banking-test';

let query = null;
const userId = '5e7abd771f06a07014d872c0';
const companyId = '5e7abc6792323e650e6c76e2';

describe('Gherkin Banking:', function() {
    beforeAll(async function() {
        await createDatabaseConnection(TEST_DB_NAME);
        const server = instantiateServer({
            context: () => ({
                userId,
                companyId
            })
        });
        ({ query } = createTestClient(server));
    });
    afterAll(async function() {
        await dropDatabase();
        await closeDatabaseConnection();
    });

    describe('Wallet:', function() {
        describe('Company wallet:', function() {
            test('Create a wallet without variables', async function() {
                const result = await query({
                    query: CREATE_WALLET
                });
                expect(result).toMatchSnapshot();
            });

            test('Create a wallet with a balance', async function() {
                const result = await query({
                    query: CREATE_WALLET,
                    variables: {
                        balance: 10.87
                    }
                });
                expect(result).toMatchSnapshot();
            });

            test('Create a wallet with a currency', async function() {
                const result = await query({
                    query: CREATE_WALLET,
                    variables: {
                        currency: 'GBP'
                    }
                });
                expect(result).toMatchSnapshot();
            });

            test('Create a wallet with a currency and a balance', async function() {
                const result = await query({
                    query: CREATE_WALLET,
                    variables: {
                        currency: 'EUR',
                        balance: 94.37
                    }
                });
                expect(result).toMatchSnapshot();
            });

            test('Fetch company wallets', async function() {
                const result = await query({
                    query: GET_WALLETS
                });

                expect(result).toMatchSnapshot();
            });
        });

        describe('Master wallet:', function() {
            test('Create a master wallet', async function() {
                const result = await query({
                    query: CREATE_WALLET,
                    variables: {
                        isMaster: true
                    }
                });
                expect(result).toMatchSnapshot();
            });

            test('Create a master wallet with a balance', async function() {
                const result = await query({
                    query: CREATE_WALLET,
                    variables: {
                        balance: 48.31,
                        isMaster: true
                    }
                });
                expect(result).toMatchSnapshot();
            });

            test('Create a master wallet with a currency', async function() {
                const result = await query({
                    query: CREATE_WALLET,
                    variables: {
                        currency: 'EUR',
                        isMaster: true
                    }
                });
                expect(result).toMatchSnapshot();
            });

            test('Create a master wallet with a currency and a balance', async function() {
                const result = await query({
                    query: CREATE_WALLET,
                    variables: {
                        currency: 'GBP',
                        balance: 78.56,
                        isMaster: true
                    }
                });
                expect(result).toMatchSnapshot();
            });
        });
    });

    describe('Card:', function() {
        test('Create a card', async function() {
            const result = await query({
                query: CREATE_CARD,
                variables: {
                    walletId: '123',
                    userId: '456',
                    balance: 0,
                    currency: 'USD'
                }
            });

            console.log(result.errors[0].extensions);
            // expect(result).toMatchSnapshot();
        });

        test('Fetch user cards', async function() {
            const result = await query({
                query: GET_CARDS
            });

            // expect(result).toMatchSnapshot();
        });
    });
});
