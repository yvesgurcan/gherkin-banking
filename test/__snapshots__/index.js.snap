// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gherkin Banking: Wallet: Company wallet: Create a wallet with a balance 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 10.87,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "USD",
      "isMaster": false,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Company wallet: Create a wallet with a currency 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 0,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "GBP",
      "isMaster": false,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Company wallet: Create a wallet with a currency and a balance 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 94.37,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "EUR",
      "isMaster": false,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Company wallet: Create a wallet without variables 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 0,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "USD",
      "isMaster": false,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Company wallet: Fetch company wallets 1`] = `
Object {
  "data": Object {
    "wallets": Array [
      Object {
        "balance": 0,
        "companyId": "5e7abc6792323e650e6c76e2",
        "currency": "USD",
        "isMaster": false,
      },
      Object {
        "balance": 10.87,
        "companyId": "5e7abc6792323e650e6c76e2",
        "currency": "USD",
        "isMaster": false,
      },
      Object {
        "balance": 0,
        "companyId": "5e7abc6792323e650e6c76e2",
        "currency": "GBP",
        "isMaster": false,
      },
      Object {
        "balance": 94.37,
        "companyId": "5e7abc6792323e650e6c76e2",
        "currency": "EUR",
        "isMaster": false,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Master wallet: Create a master wallet 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 0,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "USD",
      "isMaster": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Master wallet: Create a master wallet with a balance 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 48.31,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "USD",
      "isMaster": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Master wallet: Create a master wallet with a currency 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 0,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "EUR",
      "isMaster": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Gherkin Banking: Wallet: Master wallet: Create a master wallet with a currency and a balance 1`] = `
Object {
  "data": Object {
    "createWallet": Object {
      "balance": 78.56,
      "companyId": "5e7abc6792323e650e6c76e2",
      "currency": "GBP",
      "isMaster": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
